### PR TITLE
[Core] Unify ADAL and MSAL error handler

### PR DESCRIPTION
**Pull request by jiasli**
## Description

Refine the error message reported by #16209, #16641, #17142.

In the current code, wrapping `adal.AdalError` and rephrasing the error message actually doesn't provide more information. Instead, it leaves the original server error unexposed.

| error code | CLI error  | server error |
| - | - | - |
| `AADSTS70008` | Credentials have expired due to inactivity. | https://login.microsoftonline.com/error?code=70008 | 
| `AADSTS50079`	| Configuration of your account was changed. | https://login.microsoftonline.com/error?code=50079
| `AADSTS50173` | The credential data used by CLI has been expired because you might have changed or reset the password. | https://login.microsoftonline.com/error?code=50173

This PR 

1. wraps `adal.AdalError` in `AuthenticationError` with the original server error unchanged
2. adds `az login` instruction as `recommendation`

### Test expired refresh token

1. Create a Conditional Access policy in AAD tenant to make the refresh token only valid for 1 hour.
    image
2. `az login` with the user that is managed by the policy
3. After one hour, run `az group list` or `az account get-access-token`

## Error message

```
> az group list
AADSTS70043: The refresh token has expired or is invalid due to sign-in frequency checks by conditional access. The token was issued on 2021-02-23T05:06:57.0691981Z and the maximum allowed lifetime for this request is 3600.
Trace ID: 4999f039-c867-4cc2-ab32-7dc259780c00
Correlation ID: 6517ef89-dd65-47e1-b5fa-7193461fb78a
Timestamp: 2021-02-24 07:13:32Z
To re-authenticate, please run `az login`. If the problem persists, please contact your tenant administrator.

> az account get-access-token
AADSTS70043: The refresh token has expired or is invalid due to sign-in frequency checks by conditional access. The token was issued on 2021-02-23T05:06:57.0691981Z and the maximum allowed lifetime for this request is 3600.
Trace ID: ab764987-048b-4745-bc32-ba7d16840b00
Correlation ID: a2d45c9f-41de-4e18-8bf5-8368975c7561
Timestamp: 2021-02-24 06:54:48Z
To re-authenticate, please run `az login`. If the problem persists, please contact your tenant administrator.
```

### CODE REVIEW

**Comment by jiasli** [comment]
Exposing the original error message is the same behavior as logging in in the browser.

For example, with invalid `client_id`:

```diff
- 04b07795-8ddb-461a-bbee-02f9e1bf7b46
+ 04b07795-8ddb-461a-bbee-02f9e1bf7b41
                                     ^
```

https://login.microsoftonline.com/common/oauth2/authorize?response_type=code&client_id=04b07795-8ddb-461a-bbee-02f9e1bf7b41&redirect_uri=http://localhost:8400&state=vsjy0x6w4z3ojcnq1ezl&resource=https://management.core.windows.net/&prompt=select_account&sso_reload=true

image

**Comment by yonzhan** [comment]
Core

**Comment by jiasli** [comment]
The exposure of original AAD error paves the way for solving #16988 and #16989.

File: src/azure-cli-core/azure/cli/core/adal_authentication.py [hunk 1]
```diff
@@ -246,3 +231,27 @@ def _timestamp(dt):
     # https://docs.python.org/3/library/unittest.mock-examples.html#partial-mocking
     # https://williambert.online/2011/07/how-to-unit-testing-in-django-with-mocking-and-patching/
     return dt.timestamp()
+
+
+def aad_error_handler(error: dict):
+    """ Handle the error from AAD server returned by ADAL or MSAL. """
+    login_message = ("To re-authenticate, please {}. If the problem persists, "
+                     "please contact your tenant administrator."
+                     .format("refresh Azure Portal" if in_cloud_console() else "run `az login`"))
```
Comment (fengzhou-msft) [inline src/azure-cli-core/azure/cli/core/adal_authentication.py:240]: Should we use single-quoted `'az login'` as before?

File: src/azure-cli-core/azure/cli/core/adal_authentication.py [hunk 2]
```diff
@@ -246,3 +231,27 @@ def _timestamp(dt):
...
+
+
+def adal_error_handler(err: adal.AdalError):
+    """ Handle AdalError. """
+    try:
+        aad_error_handler(err.error_response)
+    except AttributeError:
+        # In case of AdalError created as
+        #   AdalError('More than one token matches the criteria. The result is ambiguous.')
+        raise CLIError(str(err))
```
Comment (fengzhou-msft) [inline src/azure-cli-core/azure/cli/core/adal_authentication.py:257]: Is it possible to use one of the new [Error Types](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md#available-error-types)?

File: src/azure-cli-core/azure/cli/core/_profile.py [hunk 3]
```diff
@@ -631,6 +631,10 @@ def get_msal_token(self, scopes, data):
         authority = posixpath.join(self.cli_ctx.cloud.endpoints.active_directory, tenant)
         app = ClientApplication(_CLIENT_ID, authority=authority)
         result = app.acquire_token_by_refresh_token(refresh_token, scopes, data=data)
+
+        if 'error' in result:
```
Comment (houk-ms) [inline src/azure-cli-core/azure/cli/core/_profile.py:635]: Is this a reliable way to assert the error?

File: src/azure-cli-core/azure/cli/core/adal_authentication.py (hunk 1 above)
Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/adal_authentication.py:240]: We use backticks to mark commands, like 

https://github.com/Azure/azure-cli/blob/bba13d6f698eac8eabda83575930c0ae4d399c04/src/azure-cli-core/azure/cli/core/commands/parameters.py#L241

File: src/azure-cli-core/azure/cli/core/adal_authentication.py (hunk 2 above)
Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/adal_authentication.py:257]: This error is caused by a known unknown issue in ADAL (#15320). As we are migrating to MSAL, this ADAL issue is not planned to be investigated and fixed. @houk-ms which exception do you prefer to use?

File: src/azure-cli-core/azure/cli/core/_profile.py (hunk 3 above)
Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/_profile.py:635]: Detecting `error` key in the `result` `dict` _is_ the reliable way to assert the error. It is MSAL's designed API:

https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/78e9ccfab7c16154ba38c0de78c495ba2ee58849/msal/application.py#L761-L766

```py
    def acquire_token_silent_with_error(
            self,
            ...):
        """Acquire an access token for given account, without user interaction.
        ...
        :return:
            - A dict containing no "error" key,
              and typically contains an "access_token" key,
              if cache lookup succeeded.
            - None when there is simply no token in the cache.
            - A dict containing an "error" key, when token refresh failed.
        """
```

File: src/azure-cli-core/azure/cli/core/adal_authentication.py (hunk 2 above)
Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/adal_authentication.py]: As discussed offline, raising `UnknownError` for now and also providing comment and a `recommendation`:

https://github.com/Azure/azure-cli/blob/c62df4d79ac61fc40a857897a5b6e2d83ed17830/src/azure-cli-core/azure/cli/core/adal_authentication.py#L255-L259

**Review by houk-ms — APPROVED** [review]

**Review by fengzhou-msft — APPROVED** [review]

**Pull request merged**